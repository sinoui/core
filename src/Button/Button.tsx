import React from 'react';
import BaseButton from '@sinoui/core/BaseButton';
import type { Props as BaseButtonProps } from '@sinoui/core/BaseButton';
import styled, { css } from 'styled-components';
import classNames from 'classnames';
import { opacify } from 'polished';
import getColorFromTheme from '@sinoui/core/utils/getColorFromTheme';
import colorCss from '@sinoui/core/utils/colorCss';

export interface Props extends BaseButtonProps {
  /**
   * 是否为轮廓按钮，默认为false
   */
  outlined?: boolean;
  /**
   * 是否为容器按钮，默认为false
   */
  raised?: boolean;
  /**
   * 是否禁用海拔高度，默认为false
   */
  disableElevation?: boolean;
  /**
   * 按钮颜色
   */
  color?: string;
  /**
   * 子元素
   */
  children?: React.ReactNode;
  ref?: React.Ref<HTMLButtonElement>;
}

const textButtonStyle = css<Props>`
  ${({ theme }) => ({ ...theme.typography.button })};
  min-width: 64px;
  height: 36px;
  color: ${({ theme, color, disabled }) =>
    disabled ? theme.palette.text.disabled : getColorFromTheme(theme, color)};
  user-select: none;
  transition: ${({ theme: { transitions } }) =>
    transitions.create(['color', 'background-color'], {
      easing: transitions.easing.easeInOut,
      duration: transitions.duration.shortest,
    })};
  padding: 0px 8px;
  box-sizing: border-box;

  &:hover {
    background-color: ${({ theme, color }) =>
      opacify(
        theme.palette.action.hoverOpacity - 1,
        getColorFromTheme(theme, color) || theme.palette.primary.main,
      )};

    @media (hover: none) {
      background-color: transparent;
    }
  }
`;

const outlinedStyle = css`
  border: 1px solid currentColor;
  border-radius: 4px;
  padding: 0 16px;
`;

const raisedStyle = css<Props>`
  padding: 0 16px;
  box-shadow: ${({ disabled, disableElevation, theme }) =>
    !disabled && !disableElevation && theme.shadows[2]};
  color: ${({ theme, disabled }) =>
    disabled
      ? theme.palette.text.disabled
      : theme.palette.primary.contrastText};
  background-color: ${({ theme, color, disabled }) =>
    disabled
      ? theme.palette.action.disabledBackground
      : getColorFromTheme(theme, color)};
  border-radius: 4px;
  transition: ${({ theme: { transitions } }) =>
    transitions.create(['color', 'background-color', 'box-shadow'], {
      easing: transitions.easing.easeInOut,
      duration: transitions.duration.shortest,
    })};

  &:hover {
    ${colorCss('background-color')};
    box-shadow: ${({ disableElevation, theme }) =>
      !disableElevation && theme.shadows[4]};

    @media (hover: none) {
      ${colorCss('background-color')};
    }
  }

  &:focus {
    box-shadow: ${({ disableElevation, theme }) =>
      !disableElevation && theme.shadows[4]};
  }

  &:active {
    box-shadow: ${({ disableElevation, theme }) =>
      !disableElevation && theme.shadows[8]};
  }
`;

const flowingIconStyle = css<Props>`
  margin-left: 8px;
  margin-right: ${({ outlined, raised }) => (outlined || raised ? -4 : 0)}px;
`;

const ButtonLayout = styled(BaseButton)<Props>`
  ${textButtonStyle};
  ${(props) => props.outlined && outlinedStyle};
  ${(props) => props.raised && raisedStyle};

  > .sinoui-svg-icon {
    font-size: 18px;
    height: 18px;
    width: 18px;
    margin-right: 8px;
    ${({ outlined, raised }) => (outlined || raised) && `margin-left:-4px`};
    ${(props) =>
      Array.isArray(props.children) &&
      typeof props.children[0] === 'string' &&
      flowingIconStyle};
  }
`;

/**
 * 按钮组件
 */
const Button: React.FC<Props> = React.forwardRef<HTMLButtonElement, Props>(
  (props, ref) => {
    const {
      as,
      className,
      outlined,
      raised,
      color = 'primary',
      ...rest
    } = props;
    return (
      <ButtonLayout
        {...rest}
        forwardedAs={as}
        ref={ref}
        className={classNames(
          'sinoui-button',
          {
            'sinoui-button--outlined': outlined,
            'sinoui-button--raised': raised,
          },
          className,
        )}
        color={color}
        outlined={outlined}
        raised={raised}
      />
    );
  },
);

if (process.env.NODE_ENV === 'development') {
  Button.displayName = 'Button';
}

export default Button;
