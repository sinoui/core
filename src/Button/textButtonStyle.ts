import { css } from 'styled-components';
import { opacify } from 'polished';

const textButtonStyle = css`
  ${({ theme }) => ({ ...theme.typography.button })};
  min-width: 64px;
  height: 36px;
  color: ${({ theme }) => theme.palette.primary.main};
  user-select: none;
  transition: background-color 100ms ease-in-out;
  padding: 0px 8px;
  overflow: hidden;
  box-sizing: border-box;

  &:hover {
    background-color: ${({ theme }) =>
      opacify(
        theme.palette.action.hoverOpacity - 1,
        theme.palette.primary.main,
      )};
  }
`;

export default textButtonStyle;
