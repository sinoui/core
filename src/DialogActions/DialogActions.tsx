import React from 'react';
import styled, { css } from 'styled-components';
import classNames from 'classnames';

const columnLayout = css`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin: 0px 8px 8px 0px;
  & .sinoui-button {
    margin: 0;
  }
`;

const defaultLayout = css`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  & .sinoui-button {
    margin: 8px 0;
    margin-right: 8px;
  }
`;

const DialogActionsWrapper = styled.div.attrs(({ className }) => ({
  className: classNames('sinoui-dialog-actions', className),
}))<{ column?: boolean }>`
  min-height: 52px;
  flex: 0 0 auto;
  ${(props) => (props.column ? columnLayout : defaultLayout)};
  & .sinoui-button {
    min-width: auto;
    padding: 0px 8px;
  }
`;

export interface Props {
  /**
   * 是否垂直显示
   */
  column?: boolean;
  /**
   * 子元素
   */
  children?: React.ReactNode | any;
  /**
   * 添加自定义类名
   */
  className?: string;
}

/**
 * 弹窗按钮组件
 */
function DialogActions(props: Props) {
  const { children, column } = props;
  return (
    <DialogActionsWrapper column={column}>
      {React.Children.map(
        children,
        (child: React.ReactElement<{ color?: string }>) => {
          return React.isValidElement(child)
            ? React.cloneElement(child, {
                color: child.props.color || 'primary',
              })
            : child;
        },
      )}
    </DialogActionsWrapper>
  );
}

export default DialogActions;
