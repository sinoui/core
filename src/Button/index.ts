import Button from './Button';
import type { Props } from './Button';

export type { Props };
export default Button;
