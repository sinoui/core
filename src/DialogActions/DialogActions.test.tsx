import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import DialogActions from '@sinoui/core/DialogActions';
import Button from '@sinoui/core/Button';

describe('DialogActions组件 单元测试', () => {
  afterEach(cleanup);
  test('测试按钮垂直显示', () => {
    const { container } = render(
      <ThemeProvider theme={defaultTheme}>
        <DialogActions column>
          <Button>Disagree</Button>
          <Button>Agree</Button>
        </DialogActions>
      </ThemeProvider>,
    );
    const test = container.querySelector('.sinoui-dialog-actions');
    expect(test).toHaveStyle('flex-direction:column');
  });
});

describe('DialogActions组件 快照测试', () => {
  it('基本使用', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <DialogActions>
            <Button>Disagree</Button>
            <Button>Agree</Button>
          </DialogActions>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('按钮垂直显示', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <DialogActions column>
            <Button>Disagree</Button>
            <Button>Agree</Button>
          </DialogActions>
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
