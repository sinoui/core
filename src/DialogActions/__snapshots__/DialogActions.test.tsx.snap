// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogActions组件 快照测试 基本使用 1`] = `
<div
  className="DialogActions__DialogActionsWrapper-ibc7so-0 kTvEtc sinoui-dialog-actions"
>
  <button
    className="BaseButton__BaseButtonLayout-wk8p17-0 gJsqSy sinoui-base-button Button__ButtonLayout-sc-15sa99-0 duCArv sinoui-button"
    color="primary"
    role="button"
    type="button"
  >
    Disagree
  </button>
  <button
    className="BaseButton__BaseButtonLayout-wk8p17-0 gJsqSy sinoui-base-button Button__ButtonLayout-sc-15sa99-0 duCArv sinoui-button"
    color="primary"
    role="button"
    type="button"
  >
    Agree
  </button>
</div>
`;

exports[`DialogActions组件 快照测试 按钮垂直显示 1`] = `
<div
  className="DialogActions__DialogActionsWrapper-ibc7so-0 bfxQEv sinoui-dialog-actions"
>
  <button
    className="BaseButton__BaseButtonLayout-wk8p17-0 gJsqSy sinoui-base-button Button__ButtonLayout-sc-15sa99-0 duCArv sinoui-button"
    color="primary"
    role="button"
    type="button"
  >
    Disagree
  </button>
  <button
    className="BaseButton__BaseButtonLayout-wk8p17-0 gJsqSy sinoui-base-button Button__ButtonLayout-sc-15sa99-0 duCArv sinoui-button"
    color="primary"
    role="button"
    type="button"
  >
    Agree
  </button>
</div>
`;
